### PR TITLE
Refactoring TestBackend

### PR DESCRIPTION
**Pull request by FlaminGuy**
#455

### CODE REVIEW

**Comment by FlaminGuy** [comment]
Please don't merge this until we are done.

File: main/test/ca/uwaterloo/flix/language/feature/FeatureSuite.scala [hunk 1]
```diff
@@ -28,7 +28,22 @@ class FeatureSuite extends Suites(
   new FlixTest("Test.Expression.Float32", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Float32.flix"),
   new FlixTest("Test.Expression.Tuple", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Tuple.flix"),
   new FlixTest("TestNativeCall", "main/test/ca/uwaterloo/flix/language/feature/TestNativeCall.flix"),
-  new FlixTest("TestPatternMatch", "main/test/ca/uwaterloo/flix/language/feature/TestPatternMatch.flix")
+  new FlixTest("TestPatternMatch", "main/test/ca/uwaterloo/flix/language/feature/TestPatternMatch.flix"),
+  new FlixTest("Test.Expression.Float", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Float.flix"),
+  new FlixTest("Test.Expression.Float64", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Float64.flix"),
+  new FlixTest("Test.Expression.Int", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Int.flix"),
+  new FlixTest("Test.Expression.Int8", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Int8.flix"),
+  new FlixTest("Test.Expression.Int16", "main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Int16.flix"),
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/FeatureSuite.scala]: We should keep these in order of the AST nodes, so `Test.Expression.Bool` before `Test.Expression.Float`.

File: main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Float.flix [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// TODO: More tests when we get the syntax for exponents. More tests when we have standard library (NaN, +/infinity).
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Float.flix]: I believe we have NaN and infinite in the library

File: main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Lambda.flix [hunk 3]
```diff
@@ -0,0 +1,139 @@
+/////////////////////////////////////////////////////////////////////////////
+// Lambdas - Expression.{MkClosureRef,ApplyRef,ApplyClosure}               //
+// Note that closure conversion and lambda lifting means we don't actually //
+// have lambdas in the AST. A lot of functionality is tested indirectly    //
+// by pattern matching.                                                    //
+/////////////////////////////////////////////////////////////////////////////
+
+// TODO: More tests when the typer handles lambda expressions.
+// Test actual lambda expressions (not just top-level definitions): passing them around, free variables, etc.
+// Note: this also applies to comparison (?) of lambdas and using lambdas in let-expressions.
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Lambda.flix]: We should be able to do this now. See also #477 for all the stuff we should test.

File: main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Ref.flix [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+// TODO: LoadExpression and StoreExpression tests.
+// {Load,Store}Expressions are generated, and not explicitly written in a Flix program
+
+/////////////////////////////////////////////////////////////////////////////
+// Expression.Var                                                          //
+// Tested indirectly by Expression.{Lambda,Let}.                           //
+/////////////////////////////////////////////////////////////////////////////
+
+/////////////////////////////////////////////////////////////////////////////
+// Expression.Ref                                                          //
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Ref.flix]: This should now be called `Expression.Def` and the same for the tests.

File: main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Str.flix [hunk 5]
```diff
@@ -0,0 +1,8 @@
+@test
+def testStr01(): Str = ""
+
+@test
+def testStr02(): Str = "Hello World!"
+
+@test
+def testStr03(): Str = "asdf"
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Str.flix]: I guess concatenation would also be good to test.

**Comment by magnus-madsen** [comment]
It looks close to what I want. I added a few comments. One overall comment is that we write `testChar01` or `testPlus01` but sometimes the tests are called `logicalNegateTest` so swapping the order of where the test goes in the name.

**Comment by magnus-madsen** [comment]
How about: I give some feedback on this, we fix the issues, and then we merge it in, and worry about the remaining refactorings later. Sounds good?

**Comment by FlaminGuy** [comment]
Yeah, I wanted to ask you to open refactoring tickets for todos. This just takes too much for what it is worth.

**Comment by magnus-madsen** [comment]
Nice!

**Pull request merged**